feat(share_plus): Added excludedActivityTypes support for the iOS platform.

StanleyCocos opened the pull request:
The iOS platform is missing the exclusion feature. This PR adds the platform exclusion functionality for iOS.

fix: https://github.com/fluttercommunity/plus_plugins/issues/3377

## Checklist

- [x] I read the [Contributor Guide](https://github.com/fluttercommunity/plus_plugins/blob/main/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] I titled the PR using [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0).
- [x] I did not modify the `CHANGELOG.md` nor the plugin version in `pubspec.yaml` files.
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate that with a `!` in the title as explained in [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0)).
- [x] No, this is *not* a breaking change.

StanleyCocos committed:
feat(share_plus): Added excludedActivityTypes support for the iOS platform.

miquelbeltran commented:
Thanks for this contribution as well! We have to keep this one on hold because I plan to refactor the `share()` method parameters, as also discussed in https://github.com/fluttercommunity/plus_plugins/pull/3372

Your input on this topic would be appreciated as well.

StanleyCocos commented:
@miquelbeltran 
Thank you for your reply. 
I understand your plan now. If needed, I can still improve this PR after your plan.

StanleyCocos committed:
added test parameter

StanleyCocos committed:
Merge branch 'main' into feat/add_activity_type

StanleyCocos committed:
fix(share_plus): the ambiguity caused by the parameter name.

miquelbeltran commented:
FYI, started working on the shareplus refactor, that should simplify adding this and other optional parameters: https://github.com/fluttercommunity/plus_plugins/pull/3404

StanleyCocos commented:
Hi, that's great! I've been studying the code in your branch. It's really well-organized and tidy. Next, I need to add the parameter excludedActivityType for this pull request (PR) to ShareParams. After that, I'll wait for you to finish merging xx into the main branch. Once that's done, I'll synchronize my changes accordingly.

StanleyCocos committed:
Merge branch 'main' into feat/add_activity_type

# Conflicts:
#	packages/share_plus/share_plus/ios/share_plus/Sources/share_plus/FPPSharePlusPlugin.m
#	packages/share_plus/share_plus/lib/share_plus.dart
#	packages/share_plus/share_plus/lib/src/share_plus_linux.dart
#	packages/share_plus/share_plus/lib/src/share_plus_web.dart
#	packages/share_plus/share_plus/lib/src/share_plus_windows.dart
#	packages/share_plus/share_plus_platform_interface/lib/method_channel/method_channel_share.dart
#	packages/share_plus/share_plus_platform_interface/lib/platform_interface/share_plus_platform.dart
#	packages/share_plus/share_plus_platform_interface/test/share_plus_platform_interface_test.dart

StanleyCocos committed:
add example

StanleyCocos committed:
fix dart format

StanleyCocos committed:
fix oc format

StanleyCocos committed:
++

StanleyCocos committed:
fix test

StanleyCocos committed:
fix:  format

miquelbeltran commented:
Thanks! I haven't looked too deep into it yet, but can you describe how it works in this PR, so I can understand the change better. 

And as well, add a bit of documentation about this feature on the `README.md` so users are aware of it.

StanleyCocos committed:
update readme.md

StanleyCocos commented:
Let me briefly describe the changes in this PR:

- In share_plus_platform_interface, I declared a CupertinoActivityType enum and added it as a parameter to ShareParams. Before passing it to the iOS layer, it is transmitted as a string to make it easier for users to exclude unwanted share options on iOS and macOS.

-  On the iOS native side, the string passed from the Flutter layer is received and converted back into an enum, which is then applied to the share configuration.

-  In the example, I added a multi-selection page for easier testing. Due to structural issues, I also introduced a new MyApp class that returns a MaterialApp (this caused a number of structural and formatting changes).

miquelbeltran commented on packages/share_plus/share_plus_platform_interface/test/share_plus_platform_interface_test.dart:
```diff
@@ -95,11 +97,11 @@ void main() {
-          text: 'some text to share',
-          sharePositionOrigin: const Rect.fromLTWH(1.0, 2.0, 3.0, 4.0),
-        ),
+            files: [XFile(fd.path)],
+            subject: 'some subject to share',
+            text: 'some text to share',
+            sharePositionOrigin: const Rect.fromLTWH(1.0, 2.0, 3.0, 4.0),
+            excludedActivityType: null),
```
Trailing comma missing after `null`

StanleyCocos replied:
Thank you, it’s been updated.

miquelbeltran commented on packages/share_plus/share_plus/README.md:
````diff
@@ -110,6 +110,17 @@ if (result.status == ShareResultStatus.dismissed) {
 ```
 
+On iOS or macOS, if you want to exclude certain options from appearing in your share sheet,
+you can set the excludedActivityTypes array.
+For the list of supported excludedActivityTypes, you can refer to Apple’s [documentation](https://developer.apple.com/documentation/uikit/uiactivityviewcontroller/excludedactivitytypes/).
````
```suggestion
you can set the `excludedActivityTypes` array.
For the list of supported `excludedActivityTypes`, you can refer to Apple's [documentation](https://developer.apple.com/documentation/uikit/uiactivityviewcontroller/excludedactivitytypes/).
```

Small formatting suggestion

miquelbeltran replied:
Shouldn't devs check the `CupertinoActivityType` enum instead of the Apple docs?

StanleyCocos replied:
You’re right, it should be based on CupertinoActivityType first.

miquelbeltran commented on packages/share_plus/share_plus_platform_interface/lib/platform_interface/share_plus_platform.dart:
```diff
@@ -136,6 +136,12 @@ class ShareParams {
+  ///
+  /// * Supported platforms: iOS, macOS
+  ///   Parameter ignored on other platforms.
+  final List<CupertinoActivityType>? excludedActivityType;
```
Since it's an apple exclusive param, can we name it so it references that? e.g. `excludedCupertinoActivities` since the type is `CupertinoActivityType`.

StanleyCocos replied:
Your suggestion is excellent — the naming alone makes the purpose clear to developers.

miquelbeltran commented on packages/share_plus/share_plus_platform_interface/test/share_plus_platform_interface_test.dart:
```diff
@@ -64,23 +64,25 @@ void main() {
-      ),
+          uri: Uri.parse('https://pub.dev/packages/share_plus'),
+          sharePositionOrigin: const Rect.fromLTWH(1.0, 2.0, 3.0, 4.0),
+          excludedActivityType: [CupertinoActivityType.airDrop]),
```
Missing trailing comma

StanleyCocos replied:
Thank you, it’s been updated.

miquelbeltran reviewed (commented):
Just some comments, I like the overall idea. I will also have to test this somehow.

StanleyCocos committed:
++

StanleyCocos committed:
++

miquelbeltran commented on packages/share_plus/share_plus_platform_interface/lib/platform_interface/share_plus_platform.dart:
```diff
@@ -203,3 +210,34 @@ enum ShareResultStatus {
+  addToHomeScreen,
+}
+
+extension CupertinoActivityTypeValue on CupertinoActivityType {
+  String get value => toString().split('.').last;
+}
```
What is this extension used for? I don't see it used in the PR

StanleyCocos replied:
It’s used at line 73 in method_channel_share. But I think turning it into a separate method might be a better approach now.

miquelbeltran commented on packages/share_plus/share_plus_platform_interface/lib/platform_interface/share_plus_platform.dart:
```diff
@@ -203,3 +210,34 @@ enum ShareResultStatus {
   unavailable,
 }
+
+/// An abstract class that you subclass to implement app-specific services
+/// for iOS and macOS.
+///
+/// https://developer.apple.com/documentation/uikit/uiactivity/activitytype
```
This comment doesn't seem correct to me, as these are enum values not abstract classes. Can this be reworded differently?

StanleyCocos replied:
You’re right — I may have overlooked updating the documentation here. It’s now been revised; could you please take another look?

StanleyCocos committed:
feat: Adjust the documentation comments for CupertinoActivityType, remove the extension, and add a new get value method.

StanleyCocos committed:
Merge branch 'main' into feat/add_activity_type

StanleyCocos committed:
fix: rename parameters and refactor the conversion method.

StanleyCocos committed:
fix: format

StanleyCocos commented on packages/share_plus/share_plus/ios/share_plus/Sources/share_plus/FPPSharePlusPlugin.m:
```diff
@@ -371,14 +428,26 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
+      }
+    }
+    activityViewController.excludedActivityTypes = excludedActivityTypes;
+  }
```
Thank you for your suggestion — I’ve made the adjustments. Much appreciated!

StanleyCocos committed:
fix: delete CupertinoActivityTypeValue

miquelbeltran commented on packages/share_plus/share_plus_platform_interface/lib/method_channel/method_channel_share.dart:
```diff
@@ -67,6 +67,13 @@ class MethodChannelShare extends SharePlatform {
       map['mimeTypes'] = mimeTypes;
     }
 
+    if (params.excludedCupertinoActivities != null &&
+        params.excludedCupertinoActivities!.isNotEmpty) {
+      final activityTypes =
+          params.excludedCupertinoActivities!.map((e) => e.value).toList();
+      map['excludedActivityType'] = activityTypes;
+    }
+
```
Sorry, I'll have to request another change. We renamed the variable `excludedCupertinoActivities` as it previously was `excludedActivityType`. However, I think it would be good if the map key would also be the same, to avoid confusion with that. I missed that when I first reviewed.

StanleyCocos replied:
Thank you for your suggestion. 

Slightly adjusted, but I still have some other thoughts.

I have a slightly different idea about this part. I think using excludedCupertinoActivities as the key name in the map may not be entirely appropriate, because accessing a key with “Cupertino” in its name on the iOS side feels a bit odd.

Also, I think if Android were to support a similar feature, we should still use this same key name to pass it down (even though Android doesn’t support it directly, it can potentially be handled by excluding package names).
If we do implement it for Android, I think we could introduce a new List<String> parameter for excluded package names.

miquelbeltran reviewed (changes requested):
Requesting one more change, and I think this would be completed. Thanks for keeping up with the PR!

StanleyCocos committed:
feat: rename parameter

StanleyCocos committed:
fix: test

StanleyCocos committed:
++

miquelbeltran commented:
Thanks for the changes, I think it's good to go now!
